postgres/postgis_mapmatching_plpython.md

```diff
@@ -0,0 +1,183 @@
+
+
+## Requirements
+
+- PostgreSQL with **PostGIS enabled**, we can happily recommend [Kartoza](kartoza.com)'s [docker image]([Kartoza](https://github.com/kartoza/docker-postgis)) (see [this tutorial](https://gis-ops.com/postgrest-tutorial-installation-and-setup/))
+- Python >= 3.7
+- ogr2ogr
+
+## Set up PL/Python and pyvalhalla
+
+In order to be able to use PL/Python, we need to enable it. With Kartoza's Docker image, this is as easy as executing the following statement in your database client:
+
+```sql
+CREATE EXTENSION plpython3u;
+```
+> **Note**: the 'u' at the end of `plpython3u` means untrusted, since PL/Python facilitates access to the file system from Postgres (among other dangerous shenanigans), so you might want to reconsider before using it in a production environment!
+
+
+Next, we will need to make sure that we can access pyvalhalla, so we need to install it with pip (make sure to run this from within the PostGIS container if you're using one):
+
+```sh
+python -m pip install pyvalhalla
+```
+
+
+In order to make sure we can use the Valhalla bindings from PL/Python, we quickly check Valhalla's version:
+
+```sql
+CREATE OR REPLACE FUNCTION valhalla_version()
+RETURNS TEXT
+AS
+$$
+    import valhalla
+    return valhalla.__version__
+$$ LANGUAGE plpython3u;
+
+
+SELECT valhalla_version(); -- should return something like '3.0.2'
+```
+
+Finally, we need a graph that Valhalla can route on. We will map match GPS signals recorded in the Spanish region of Asturias, so one option is to download the respective OSM file from Geofabrik, and use our [Valhalla Docker image](https://hub.docker.com/r/gisops/valhalla) to build a graph from it. We have described how to use it [in another tutorial](https://gis-ops.com/valhalla-how-to-run-with-docker-on-ubuntu/), so we won't go into further detail here. For the sake of ease however, we have already added the graph tiles needed to follow this tutorial [here](https://github.com/gis-ops/tutorials/raw/master/postgres/data/valhalla_tiles.tar.gz), so don't worry if you're not familiar with Docker. Simply download the data and extract them using e.g. gzip:
+
+```sh
+gzip -d valhalla_tiles.tar.gz
+```
```

Comment:
the assumption should generally be with our hosted tutorials that people won't `git clone` the repo. so they need a way to download the file as well.

`wget https://github.com/gis-ops/tutorials/blob/27c1b30cc59702a4d87a093aa0921d155bd17f1d/postgres/data/valhalla_tiles.tar.gz` should do it